backend/session/session.c

```diff
@@ -1,32 +1,131 @@
 #define _POSIX_C_SOURCE 200809L
 #include <assert.h>
 #include <libudev.h>
+#include <stdarg.h>
 #include <stddef.h>
 #include <stdlib.h>
+#include <stdio.h>
 #include <string.h>
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <unistd.h>
 #include <time.h>
 #include <wayland-server-core.h>
 #include <wlr/backend/session.h>
-#include <wlr/backend/session/interface.h>
 #include <wlr/config.h>
 #include <wlr/util/log.h>
 #include <xf86drm.h>
 #include <xf86drmMode.h>
 #include "backend/session/session.h"
 #include "util/signal.h"
 
+#include <libseat.h>
+
 #define WAIT_GPU_TIMEOUT 10000 // ms
 
-extern const struct session_impl session_libseat;
+static void handle_enable_seat(struct libseat *seat, void *data) {
+	struct wlr_session *session = data;
+	session->active = true;
+	wlr_signal_emit_safe(&session->events.active, NULL);
+}
+
+static void handle_disable_seat(struct libseat *seat, void *data) {
+	struct wlr_session *session = data;
+	session->active = false;
+	wlr_signal_emit_safe(&session->events.active, NULL);
+	libseat_disable_seat(session->seat_handle);
+}
+
+static int libseat_event(int fd, uint32_t mask, void *data) {
+	struct libseat *seat = data;
+	if (libseat_dispatch(seat, 0) == -1) {
+		wlr_log_errno(WLR_ERROR, "Failed to dispatch libseat");
+		exit(-1);
```

Comment:
Since wlroots is a library, I'd prefer to not exit here. Although I'm not sure we can do a lot more than just logging an error.

Also see TODOs in `handle_libinput_readable` for instance.

Reply:
Indeed, it's not nice to exit. Didn't know what to do here. Just got carried over from the libweston MR.

I'll try to copy the display termination.

Reply:
We now call `wl_display_terminate`.

Reply:
Handled the rest in https://github.com/swaywm/wlroots/pull/2857